### ISSUE TITLE
make moment mostly immutable

### ISSUE THREAD

**Issue by ichernev** [moment/moment#1754]
There has been [a lot of discussions](https://github.com/moment/moment/search?q=immutable&ref=cmdform&type=Issues) about this. Here's the proposal:

The following mutable methods will become immutable in 3.0.0: `utc`, `local`, `zone`, `add`, `subtract`, `startOf`, `endOf`, `lang`, also in `duration`: `add`, `subtract`, and `lang`.

As a start, all methods will be duplicated with `methodNameMute` variants. We also need immutable variants named `methodNameImmute`. From 3.0 the plain old methods will start using the immutable option by default.

What is debatable is:
- should `lang` be made immutable
- should all the getters/setters (including `get`/`set`) be made immutable too
- naming of the mutable and immutable versions of the methods
- and of course -- should we do the switch, or just stop at the immutable API

The good part is we can make immutable versions of the methods today and decide later on what to do. If we switch it would also mean the `2.x` branch would be around for quite some time after `3.x` is out.

@icambron @timrwood @gregwebs @yang @lfnavess @soswow @langalex

**Comment by jasonhansel** [comment]
+1 for immutability

**Comment by eliseumds** [comment]
+1 for immutability. What about the ```format``` method that also mutates the object?

**Comment by Karl-Gustav** [comment]
+1 for an immutable moment.js

**Comment by boxxxie** [comment]
:100: 
do the imootables!

**Comment by SystemParadox** [comment]
+1 these days immutability is something I expect in any good JS API

**Comment by abergs** [comment]
:+1: Yes please, we would very much like this. Currently we sprinkle .clone() everywhere.

**Comment by manuelbernhardt** [comment]
+1 this would be a very nice improvement

**Comment by martiansnoop** [comment]
:+1:  immutable all the things

**Comment by adamhutch** [comment]
+1 for immutable

**Comment by bastienJS** [comment]
Make it all immutable. I am sick of cloning every moment(now) variable before I operate on now variable and then now is changed again.

**Comment by droberts84** [comment]
+1 for immutability

I would not have expected `startOf('day')` to mutate (although I will admit its well documented if I had read more carefully).  That caused a fun bug in my application.

**Comment by jdurand** [comment]
I definitely agree that most of moment.js operators are awkward in their mutability. `mmnt.startOf('day')`is super counter intuitive in that is mutates `mmnt`.

I use moment.js for calendar type use cases with a lot of loops and date comparisons. I have hit the performance issues with `clone()` and they are atrocious. Having **some** control over what clones and what mutates is essential to me and probably others.
Event though having `clone()`everywhere seems awkward at first, what it does is crystal clear and made refactoring for performance super easy for me.

If every method starts using `clone()` internally for the sake of having a nicer API, I think we'll have missed the point.

**Comment by dsherret** [comment]
@jdurand I would rather explicitly mutate than explicitly clone.

**Comment by jdurand** [comment]
@dsherret I don't mind explicitly anything. My main concern is about **not implicitly cloning** as it is a very costly operation.
Someone mentioned that setters would need to return a cloned copy and that scared me; It would be very inefficient.

**Comment by dsherret** [comment]
@jdurand it might be more inefficient, but I think an implicit clone would benefit the majority of applications where the difference between cloned and mutated objects would make no noticeable difference to the end user's experience. I think developer ease and trying to prevent developer error should take precedence over a few saved milliseconds, as the majority of developers aren't doing thousands of date operations at once. For those who are, they could maybe explicitly say they want to mutate instead of cloning.

Sidenote: with this change I think some performance improvements could be made though... for example, cloning could be eliminated by instead referencing the past immutable object and then store the operation to be done for this object (ex. `add(1, 'days')`). The result would then only be computed by executing the operations when something like `.toString()`, `.format()`, `.toDate()`, `.day()`, etc... is called. That would be a dramatic change and it might not end up being faster... some unit tests would need to be done to compare the performance (additionally there could be other issues I haven't considered as I've never looked at any code in moment.js other than how it clones).

**Comment by jdurand** [comment]
@dsherret I like the *builder*/*lazy* approach; with hindsight, it should probably have been built like this from the start.
As you said, I think an immutable-ish fork with API compatibility in mind would be the best.

**Comment by AlexGalays** [comment]
Since when are we so afraid of breaking changes? The current version is stable and people can still use it without refactoring their code base.

Maintaining two codebases is a pain, slows you down and isn't really worth it just to have a mutable/immutable version.

So let's just go with a fully immutable moment, bump the major version and be done with it :dancers:

**Comment by kevinhooke** [comment]
Found this thread as I just started using this library but have been pulling my hair out debugging some code and dealing with what to me seems to be unpredictable side effects of being mutable. Would love to see a fully immutable version of the library!

**Comment by butterflyhug** [comment]
Shameless plug:  I got tired of waiting for something to happen here.  Over the last few days, I've resurrected [Frozen Moment](https://github.com/WhoopInc/frozen-moment) and re-written it to act as a plugin for Moment itself.  Hat tip to @wyantb for helping me get the first preview release done over the weekend.

Frozen Moment provides an immutable type that works just like Moment.  Basically, the immutable version wraps Moment's functionality and calls `.clone()` for you whenever appropriate.

For those who like builder APIs, I'd challenge you to think about Moment itself as a very nice implementation of a builder object!  Frozen Moment adds the core immutable type that we all want, and a mechanism for building an immutable Frozen Moment from a mutable Moment.

For those who just want to work with a convenient immutable API -- well, I intend to support that, too.  I haven't built a constructor that will directly build a Frozen instance yet, but that's [on the TODO list](https://github.com/WhoopInc/frozen-moment/issues/4).  In the short term, the workaround is to create everything with `moment().freeze()` or `moment.utc().freeze()`.

Frozen Moment is obviously a young codebase, so there are probably a few rough edges -- but I'd encourage anyone here to give it a try and file issues for anything that doesn't work the way you'd expect.

Oh, one more thing:  I'm not advertising this yet, but Frozen Moment instances should "just work" with most Moment plugins.  Just make sure that all your other Moment plugins register themselves before Frozen Moment.  If you find a plugin that doesn't work as expected with immutable Frozen Moments, file a bug and I'll look into it.

**Comment by sandstrom** [comment]
+1 on immutability

**Comment by lee101** [comment]
+1 for immutable

**Comment by wmertens** [comment]
Did anybody look at implementing moment on top of [Immutable JS](https://facebook.github.io/immutable-js/)? It's an optimized library for immutability in JS, it reuses unchanged parts of an immutable object, reducing memory concerns.

**Comment by jbest84** [comment]
+1 
I just fixed a 3 year old bug due to this: https://github.com/Saleslogix/argos-saleslogix/commit/87b35b22c1a42670da369701b15e2bdb3786cabc

**Comment by AlexGalays** [comment]
So the users ask for immutability but the core team find excuses not to do it ? :p
Come on guys, this change is much more important than rewriting the code in ES6 ^^ In its current form, the API is simply bad, a bit like the JS Array one where some methods are immutable (filter, concat, etc) but some others are not (reverse, sort) except their constraint for backward compatibility is infinitely higher than for a library.

**Comment by faebser** [comment]
+1

**Comment by faebser** [comment]
a little trap that always gets me (and IMHO this is good reason for immutability):

```javascript
var today = moment();
for (var i = 0; i < 7; i++) {
   week.push(today.subtract(i, 'days').format('dd').toUpperCase());
}
```
this sadly does not generate a array with the day names but something strange because you you actually compute the date like this:
```javascript
i = 0 = today - 0;
i = 1 = today -0 -1;
i = 2 = today -0 -1 -2;
etc
```
so you have to refactor it into this:
```javascript
var today = moment();
for (var i = 0; i < 7; i++) {
            if (i == 0) {
                week.push(today.subtract(0, 'days').format('dd').toUpperCase());
            }
            else {
                week.push(today.subtract(1, 'days').format('dd').toUpperCase());
            }
        }
```

**Comment by SystemParadox** [comment]
@faebser excellent example

**Comment by mrcnkoba** [comment]
+1 for immutability

**Comment by webPapaya** [comment]
+1

**Comment by Wirwing** [comment]
@faebser happened to me this morning. Angular's two-way binding + mutability makes a pain in the **s to keep cloning dates instead of modifying the current ones.

**Comment by spovich** [comment]
+1 for immutability, this just cost me a couple hours.

**Comment by rhys-vdw** [comment]
:+1:

**Comment by davehughes** [comment]
+1 for immutability

**Comment by thomasvanlankveld** [comment]
I'm a little torn about this topic.

The purist in me wants to shout out "+1 for immutability! Moment objects are clearly of the [ValueObject](http://martinfowler.com/bliki/ValueObject.html) kind."

However, we can't ignore that moment.js is the [13th](https://github.com/search?l=JavaScript&p=2&q=stars%3A%3E1&s=stars&type=Repositories) most popular javascript repo on GitHub (and [24th](https://github.com/search?p=3&q=stars%3A%3E1&s=stars&type=Repositories) all-around), and that searching for "moment" on bower.io returns [111](http://bower.io/search/?q=moment) matched results. Even if we can provide application builders with a gradual way to implement an immutable version of moment, it's going to cause a huge mess among its dependencies.

@ichernev Perhaps a more humble suggestion: Would it be possible devote a small part of moment's documentation page to talking about mutability vs immutability? You could briefly state your official stance on the topic, and maybe add a brief summary of arguments, place a link to this thread, or if you want specific discussion input it could work like a CTA.

Right now the documentation page doesn't mention the term "immutable". Googling "moment immutable" takes you to this page, but it took me two hours reading and I'm still not sure about your current stance on the topic. It'd be great if Google's top hit for "moment immutable" gave a quick answer on the future of immutability in moment :)

To quote @jehoshua02:
"I sympathize that this is a hard decision to make. I appreciate your work. Moment.js is amazing."

**Comment by pitaj** [comment]
Here's my proposal. Instead of making the base moment immutable, add a `moment.immutable` factory. It would be a property of the `moment` function and would enclose the same exact API signature as `moment`, but immutable. 

It could even just be a prototype extension of the mutable `moment` factory, using the functions of that prototype, but cloning instead.

**Comment by danpantry** [comment]
@thomasvanlankveld breaking changes are what major version numbers are for. Anyone using Bower and npm will have the option of sticking to the current major version number; we should not worry about backwards compatibility here - except maybe those people who are just serving this from CDN. But if they are using momentjs from CDN they likely aren't all that interested in updating the library from time to time anyway

**Comment by KrunchMuffin** [comment]
So looks like I just ran into this issue as well.

http://stackoverflow.com/questions/33002430/moment-js-formatting-incorrect-date

So I am all for immutability across the board.

**Comment by scandox** [comment]
+1 For immutability

**Comment by RobertMcCarter** [comment]
Just lost an evening to this surprising and unexpected behaviour.
Another +1 for immutability with a corresponding major version change to make it clear.

**Comment by zbarnett** [comment]
+1 For immutability

**Comment by hanbzu** [comment]
+1 Yes please.
The collective amount of hours lost due to this must be huge.

**Comment by magwo** [comment]
Yeah I mean seriously. Who cares about performance in a datetime library? Like.. really? I'm guessing 99.9% of users aren't doing anything that even remotely requires good performance. Typically you handle a couple of dates, or in worst case a few hundred. The few users who are handling millions of dates per second can use optimised mutable API points.

Immutability is the only sane design choice. There are several studies that show that programming with immutable types is much less bug prone than with mutable types.

**Comment by sixsided** [comment]
+1 for immutability. This cost me a couple hours.

Part of the problem is that method names like `.startOf()` don't imply mutation of the underlying object.

**Comment by schmod** [comment]
I've hit performance bottlenecks with Moment, so I can assure you that this can sometimes happen.

However, I'm not convinced that immutable moments would be inherently less efficient, and can envision a few cases where they would be more efficient.

This debate was settled a long time ago in the Java world.  Immutable dates won, and the most popular implementation (JodaTime) eventually became part of the standard library in Java 8.  

Working with Java 8's `LocalTime` has been one of those slap-your-forehead "Why didn't we *always* do this?" moments.  I rarely evangelize technologies, but I honestly can't see any upsides to mutable date objects.

So, yeah.... I hate that these threads get flooded with +1s, but the truth of the matter is that somebody else is going to create an immutable JS date library if Moment doesn't.

**Comment by schmod** [comment]
I recently stumbled upon a new npm module that purports to port much of the JodaTime API (ie. Java 8 dates) to JS.

This would bring things like immutability, LocalDate, and LocalTime to node and the browser.  

Having worked with these concepts in Java, everything else feels kludge-y and bug-prone.

**Comment by wmertens** [comment]
Link?

On Fri, Dec 11, 2015, 4:30 PM Andrew Schmadel <notifications@github.com>
wrote:

> I recently stumbled upon a new npm module that purports to port much of
> the JodaTime API (ie. Java 8 dates) to JS.
>
> This would bring things like immutability, LocalDate, and LocalTime to
> node and the browser.
>
> Having worked with these concepts in Java, everything else feels kludge-y
> and bug-prone.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/moment/moment/issues/1754#issuecomment-163964349>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by mj1856** [comment]
Since I haven't chimed in yet, I'll just state that I am in favor of immutability in moment 3.0.  Primarily because I come from the DDD school of thought, where objects like `moment` would be considered *value objects*, and thus are best implemented as immutable.

Even if there's a significant performance hit, it's still the right thing to do.  Moment should fit naturally into others' designs.  Intuitive API trumps perf, and mutation is not intuitive on value objects (IMHO).

I also think moment 3.0 should remove its dependency on the `Date` object, but that's a discussion for a different thread.

**Comment by kruncher** [comment]
I have been using `Object.freeze` on moment instances and this has generally achieved what I have needed; except I just discovered that the following fails:
```javascript
let now = Object.freeze(moment());
if (now.isSameOrBefore(anotherTime)) { // throws exception
}
```

The exception:
```
     TypeError: Can't add property _isValid, object is not extensible
      at valid__isValid (C:\git\quick-test\node_modules\moment\moment.js:93:24)
      at Moment.moment_valid__isValid [as isValid] (C:\git\quick-test\node_modules\moment\moment.js:2195:16)
      at Moment.isSame (C:\git\quick-test\node_modules\moment\moment.js:1945:44)
      at Moment.isSameOrBefore (C:\git\quick-test\node_modules\moment\moment.js:1962:21)
```

Can this be fixed so that `Object.freeze` can be used when desired?

**Comment by gabrielmaldi** [comment]
@wmertens I'd assume this is it: https://github.com/pithu/js-joda

**Comment by timrwood** [comment]
@ichernev, @mj1856, as I've been haven't been involved in moment core development for a while and immutability has fairly significant interest, I'm retracting my previous stance.

I'm not sure if I was the only blocker on this, but I am comfortable moving forward with immutability in 3.0.

**Comment by schmod** [comment]
@gabrielmaldi @wmertens Yep. That was it.  Apologies for my borderline-incoherent comment -- I clearly clicked the 'Submit' button on a halfway-written post.  

To gather up some of my disjointed thoughts:

* There is clearly some interest in immutable date objects for JS.  There are mature immutable date libraries in several other languages, and there's a lot of general momentum toward immutable objects in JS ([immutable.js](https://facebook.github.io/immutable-js/) has 10,500 stars if that's any indication).  At the very least, I think this deserves a proof-of-concept. 

* In spite of this interest, very little code appears to have been written.  [js-joda](https://github.com/pithu/js-joda) appears to be the first serious attempt to write an immutable date library for JS.

* Immutable moments would be a huge breaking change, which raises some philosophical questions.  While I would hate to lose the support of the very large MomentJS community, it wouldn't necessarily be a horrible thing for those of us who are interested in immutable JS dates to make a clean break, and make an effort to contribute to js-joda instead of trying to push a fairly radical change on Moment, which is a mature library with a large and established user-base.  

* Aside: js-joda is still very young, and it's unclear what the author's goals and intentions are for the library.  In particular, he needs to pick a license, and we may want to consider whether the needs of the typical JS developer would be met by a faithful re-implementation of Joda Time or JSR-310.

**Comment by tacomanator** [comment]
+1 for an immutability, and the saner code that will result.

It will be a major effort, so a sincerest thank you to those who'll be (and have been) making it happen.

**Comment by wyantb** [comment]
Moment is widely used enough that I think it'd be actionable to go with something approaching the following, assuming the is implemented similarly to @butterflyhug's https://github.com/WhoopInc/frozen-moment 

3.x: immutable as an option, defaulted to false, and have a flag set on the global moment export that would set to true; console.warn on library load (in dev mode)
4.x: immutable as an option, defaulted to true, flag still available to be set as false; console.warn about schedule for 5.x (in dev mode)
5.x: immutable as the only way

Have a page up-front and center describing the long-term immutability vision - say, one major release a year along the 3.x/4.x/5.x outline I laid out - and I think it'd give a reasonable amount for anyone affected to update their code.

**Comment by butterflyhug** [comment]
A few observations:

1. I built [WhoopInc/frozen-moment](https://github.com/WhoopInc/frozen-moment) hoping that a bunch of folks here might be willing to build things with it, as a way to find problem points in the moment ecosystem where immutability would break plugins, etc.  So far, few people have done that, which has made me less eager to work on frozen-moment and/or community evangelism to help plugin authors support immutable moments.

  That said, I'm still willing to help fix bugs that people find when using frozen-moment -- regardless of whether those bugs are mine or whether they're in another moment plugin that never considered the possibility that moments might become immutable.  Just file issues at frozen-moment and I'll take a look.  And I still think that a community effort around something like frozen-moment could help us understand and ease the pain of a transition to immutable moments.

2. If moment implements immutability in 3.0, it would be pretty straightforward for someone to write and maintain a wrapper to implement the mutable moment 2.x API on top of the immutable 3.x API.  Conceptually, this would actually look a lot like frozen-moment:  everywhere that frozen-moment implicitly `clone()`s, this mutability wrapper would instead implicitly change its internal reference to a new immutable moment value.  This could also help ease the transition for folks using moment in large codebases.

3. I would be willing to help think through potential issues and implement immutability in moment 3, if desired.

4. js-joda will be [directly porting JSR-310](https://github.com/pithu/js-joda/issues/13#issuecomment-171216489) into JavaScript [using a BSD license](https://github.com/pithu/js-joda/issues/10#issuecomment-170118536).

**Comment by RobertMcCarter** [comment]
@butterflyhug - Thanks for building frozen-moment!  I was excited when I found that, but worried about introducing a dependency from my project to frozen-moment, given that if support was dropped changing from an immutable moment to a mutable one would be a big job in my code.   If your goal was to get feedback and you're actively supporting it then I'm more comfortable doing that.  :-)

I don't know how many other people might have had the same thought processes.

**Comment by schmod** [comment]
I'm curious if immutability should be explored in conjunction with the notion of breaking the dependency on the `Date` object (or even adding in some sort of lazy evaluation) for performance reasons.

**Comment by danpantry** [comment]
I would suggest taking things one step at a time, @schmod. This is already a (apparently) big change

**Comment by butterflyhug** [comment]
@RobertMcCarter Yep, I am planning to support frozen-moment for the foreseeable future, unless/until some option for immutability gets implemented in the core library -- in large part because I'm personally using it for some things that I expect to be maintaining for a while.  That said, my use cases don't span the full Moment ecosystem, so I rely on feedback and usage from other folks to help identify what's important.

**Comment by mindsubversion** [comment]
+1 for immutability, after half an hour spent figuring out what happened with my app after first using endOf(). Granted I didn't read the docs carefully, I just assumed a method with that name would return the endOf the month for example, and leave the moment instance unaffected. To be honest its a surprise and I find it ridiculous that it doesn't happen this way or maybe my head is too used to the -to my opinion- enormous benefits of having a -mostly- immutable API

**Comment by travisbloom** [comment]
+1 for immutability :)

**Comment by es6Test** [comment]
I don't care about immutability, leave this excellent libary alone!

**Comment by danpantry** [comment]
@es6Test On what grounds do you disagree with immutability? Do you have any concrete reason other than resistance to change?

**Comment by strokyl** [comment]
+1
I think it will ease the use of the library if it was immutable my code is full of .clone() method and sometime mutability provoke very hard to find bugs.

**Comment by ChelseaLura** [comment]
+1 Please, more immutability ^_^

**Comment by es6Test** [comment]
@danpantry After making some more apps, I've changed my mind. I would prefer immutability.

**Comment by es6Test** [comment]
But I would like to still have some vars mutable, I quite often like to add days to moments and I don't want to have to create more vars.

**Comment by danpantry** [comment]
@es6Test Just use `let` if you *really* dislike it?

````javascript
let date = moment()
// with immutability
date = date.add(5)
````

**Comment by w0rp** [comment]
+1

I'm in favour of immutability as well. That's the one feature of Python's `datetime` library which really makes it shine. You can throw objects around without worrying about something modifying them, just like with character strings.

**Comment by ericlau-solid** [comment]
In React / Redux world, immutability is super paramount. I ran into some gnarly issues already because moment isn't intrinsically immutable. Performance is irrelevant here since it can be mitigated. ImmutableJs library from Facebook proves that immutable can be done without sacrificing performance.

I also will +1 on immutability all the way. You don't get any benefit of immutability until then, so there's no point in doing in half way.

Immutability and functional programming are being used more now than ever before (and I love it). I will offer my own time to contribute to this initiative. Contributors please contact me and let me know how I can be of help.

**Comment by maggiepint** [comment]
The team wanted to give an update on this since we have had so much interest.

At this time, we want you to know that we have heard your concerns. The entire maintainers team agrees that were we writing the library from the ground up, we would choose to make it immutable. However, that is not how it is right now.

As it stands, we have 4 million NPM downloads a month, along with an uncountable number of users relying on getting moment through other channels. All of these users have code that expects moment to be mutable as it is today. In addition, we have many plugins relying on moment being mutable.

The code is easy enough to write, but the support needs of this sort of change are difficult to take on as a small team doing this in our spare time. That said, we are willing to consider it. We are wondering though: 
- Why doesn't the frozen moment plugin meet your needs?
- Do you just need an immutable API (an API that always returns a clone), or do we for some reason need to internally truly not mutate the objects?
- Would adding a second, immutable API to the current code meet your needs?

**Comment by ericlau-solid** [comment]
The issue with frozen moment plugin is that it's opt-in, not opt-out. Every
moment() needs a .freeze() for it to be immutable.

The whole point of immutability is that when I have object X, I know that
unless I explicitly make it mutable, it needs to be impervious to change.
Any public function that lets me change an object internally without
explicit mutability is problematic.

I think one workaround is to do several things:
1. Add a second immutable API to the current code
2. Have a global moment() setting where if it's set to immutable mode, all
moment() instances created are immutable, if a mutable call is attempted,
will not mutate and instead error with appropriate message to use immutable
api calls.

That I think will satisfy everyone. What do you think?



On Mon, May 23, 2016 at 12:11 PM, Maggie Pint <notifications@github.com>
wrote:

> The team wanted to give an update on this since we have had so much
> interest.
>
> At this time, we want you to know that we have heard your concerns. The
> entire maintainers team agrees that were we writing the library from the
> ground up, we would choose to make it immutable. However, that is not how
> it is right now.
>
> As it stands, we have 4 million NPM downloads a month, along with an
> uncountable number of users relying on getting moment through other
> channels. All of these users have code that expects moment to be mutable as
> it is today. In addition, we have many plugins relying on moment being
> mutable.
>
> The code is easy enough to write, but the support needs of this sort of
> change are difficult to take on as a small team doing this in our spare
> time. That said, we are willing to consider it. We are wondering though:
>
>    - Why doesn't the frozen moment plugin meet your needs?
>    - Do you just need an immutable API (an API that always returns a
>    clone), or do we for some reason need to internally truly not mutate the
>    objects?
>    - Would adding a second, immutable API to the current code meet your
>    needs?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/moment/moment/issues/1754#issuecomment-221062274>
>



-- 

Eric Lau from Gmail account.

**Comment by tacomanator** [comment]
Thanks for responding @maggiepint.

1. A few issues with frozen moment plugin: it's not the default, it requires extra work to use, there's less chance that it will be actively maintained, my assumption is that it has a performance hit (probably a very small one) over something that was built as immutable. I think the biggest problem is forgetting to use it, especially on larger projects with many engineers.

2. The API should be side-effect free. Returning a clone is good but if you also modify the original object that would be a side-effect. The following code should in not modify start when declaring end: 

```javascript
start = moment();
end = start.add(10, 'minutes');
```

3. Do you mean having functions like 'immutableAdd' in addition to 'add'? If so then yes technically it would, especially if you create a wrapper to use immutable functions using the same names:

```javascript
import "moment/immutable";
start = moment();
end = start.add(10, 'minutes'); // immutable version of add
```

In my opinion this approach lends it self to a graceful upgrade for those who want to use the immutable API—existing code will still work, nobody is forced in, and requires less changes for those who do want to use the immutable API.

**Comment by maggiepint** [comment]
We have every intention of making the code side effect free. The basic pattern would be to hook into the current functions, have them call clone, and then operate on that clone.

Technically though, if you're making truly immutable code, the objects should be constructed with values that won't change. Doing so would increase the difficulty in making this change though. It's much easier to call clone and then do what we were doing before.

**Comment by ericlau-solid** [comment]
I second Drew's idea:

import "moment/immutable";
start = moment();
end = start.add(10, 'minutes'); // immutable version of add


That would be great since it's immutable opt-out and no changes to
function names.


On Mon, May 23, 2016 at 12:53 PM, Maggie Pint <notifications@github.com>
wrote:

> We have every intention of making the code side effect free. The basic
> pattern would be to hook into the current functions, have them call clone,
> and then operate on that clone.
>
> Technically though, if you're making truly immutable code, the objects
> should be constructed with values that won't change. Doing so would
> increase the difficulty in making this change though. It's much easier to
> call clone and then do what we were doing before.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/moment/moment/issues/1754#issuecomment-221076796>
>



-- 

Eric Lau from Gmail account.

**Comment by ichernev** [comment]
@ericlau-solid
> 2. Have a global moment() setting where if it's set to immutable mode, all
moment() instances created are immutable, if a mutable call is attempted,
will not mutate and instead error with appropriate message to use immutable
api calls.

This will break if 2 libraries on the page expect different behaviors of moment. For example you use datepicker, which expects a mutable moment, and your own code expects immutable one -- it won't work.

> Having immutable api (ala frozen-moment)
This also is not perfect, because no moment plugins will work with it. Actually, there is no way to introduce immutability AND keeping all existing plugins functional. The plugins will either not work at all, or not work with immutable moments, before someone spends the time to make them work.

**Comment by ichernev** [comment]
> import "moment/immutable";

So this is basically proposing a different library with a similar interface. Sure, it won't break code but it will likely cause confusion and make some project have the 2 versions installed (for example datepicker will take the mutable version, and you code will take the immutable one).

**Comment by w0rp** [comment]
The best option is to pretty much make an API like the current one, but with immutable objects. A fork might be in order. That is the way of free software after all.

**Comment by schmod** [comment]
The "add a second API / publish a submodule" option would be best for compatibility, because developers would be able to upgrade their code incrementally.  If at all possible, the meaning of `moment()` (via `import moment;` in module-land, or the `moment` global in the browser build) should not change at all, because there's a ton of legacy code out there.

IMO, a solution that allowed the following would be ideal:

```
import moment from 'moment';
import {immutable as immoment} from 'moment';

var a = moment(); // mutable moment
var b = moment().immutable(); // immutable moment
var c = immoment(); // also an immutable moment; shorthand
```

Not a different library, but yes -- have two separate and distinct types of moment.  Assuming that the mutable and immutable moment objects share a significant amount of code "under the hood," the library size wouldn't need to increase much for developers who need to use both syntaxes.  If an application only uses one syntax, [tree-shaking](https://rollupjs.org) could be used to produce an optimized build.

Forking the library would potentially double the size of the code most developers will need to load, which would be undesirable for most web applications.  

---

I don't think that it would be a good idea to assume that any plugins will work with immutable moments "out of the box," and wouldn't consider that to be a reasonable requirement.  

I also don't think it's a reasonable requirement/assumption for the mutable and immutable APIs to be identical. We should strive to make the upgrade path as smooth as possible, but shouldn't lock ourselves into any awkward syntaxes.  

---

I also think there's a lot of room for debate/argument about what the "immutable" API would look like.  Builder pattern?  Constructor instantiation?  Value object?  Implement the current API, but have any "setter" methods return a clone?

My one request is that the new API should be clear and unambiguous about which kind of moment we're working with.  IMO, `freeze` fails that test, and offers little protection beyond calling `clone()` every time I pass a moment around.

**Comment by tacomanator** [comment]
> So this is basically proposing a different library with a similar interface. Sure, it won't break code but it will likely cause confusion and make some project have the 2 versions installed (for example datepicker will take the mutable version, and you code will take the immutable one).

Not a separate library, just a slim wrapper around the same library. There would be only need be one version of moment installed. You may import 'moment/immutable' and datepicker may import 'moment' in the same project without issue.

There could be problem in interfacing with plugins that take in or return a moment if they have not updated to handle immutable moments.

Say you you pass an immutable moment to a plugin which, at present, treats it as mutable. Until the plugin is updated, there would need to be a way to convert to mutable before a moment is given to the plugin. Ideally it would be deprecated after project maintainers have some time to bake in support for immutability. There would also probably need to be a way to detect immutability (for plugin authors).

On the flip side a library that updates to use the immutable interface, and returns an immutable moment to users, may throw off those who don't use immutability and aren't expecting an immutable moment. I believe this can also be handled by a de-be-deprecated method for conversion. Plugins should return a moment of the same type as is passed in.

**Comment by tacomanator** [comment]
@maggiepint yes I understand your point but I think that might be too big of a leap to make. Personally I would be ok with an approach that simply `clone`'s behind the curtain. Especially while the kinks are being worked out.

**Comment by maggiepint** [comment]
@tacomanator that's what I want to do as well. Redoing the whole thing to be truly immutable internally is not terrible tenable. 
I just didn't know if for some reason people WANTED that though.

**Comment by butterflyhug** [comment]
@schmod Could you briefly elaborate on why you don't think `freeze` is a clear and unambiguous API?

If the function name is too cute, I'd be happy to consider an argument for renaming Frozen Moment's methods as an issue on that repo.  On the other hand, if you don't think it should be possible to convert from mutable to immutable (and back), then I suspect that might prompt some discussion here -- although note that I'd also be willing to take Frozen Moment in that direction if we get a rough consensus that that'd be a better API.

**Comment by butterflyhug** [comment]
@maggiepint Am I correct in reading "add a second, immutable API to the current code" as basically bundling part or all of Frozen Moment with the core library?  If it'd be helpful, I'd be happy to help migrate Frozen Moment or some of its ideas into the moment organization (either as an official plugin or as part of the core library) and help with its maintenance in that new context.

**Comment by maggiepint** [comment]
@butterflyhug  You are saying exactly what @ichernev was thinking. He was leaning toward official plugin that shipped with the library. Maybe we could coordinate over this in Gitter at some point? I have to be away for three days on business, so I'll have limited availability. Rest of the guys should be around.

**Comment by danpantry** [comment]
@butterflyhug 

I think that `freeze` is a fine name, it just sucks to have to remember to use it after *every single* invocation of moment.

I also think that `import 'moment/immutable'` has the same issues, it's easy to forget to it.

Honestly, you're following semver, I think the correct approach is to make everything use immutable *by default* and release this change as it's own major version, with a plan to no longer support older versions after a time period. New features / fixes could be merged into both tracks, immutable and mutable, with a migration plan given to users using the mutable version.

Of course, this is a lot of overhead, but I think it would be better to do this change right than try and fudge it in the interests of backwards compatibility. It also makes the change a lot more complicated which may dissuade interest from the core development team. I dunno.

**Comment by magwo** [comment]
My vote is also on semver and immutable 3.0 release.

**Comment by schmod** [comment]
I don't object to the `freeze` name, but find it inherently problematic that it's difficult to make safe assumptions about whether or not a moment is frozen.  (`Object.freeze()` has a similar issue)

An immutable moment should not be un-freezable, and there should never be any doubt about whether or not it is frozen.  This could be accomplished via a strict variant of the builder pattern (ie. the object _only_ has setters until `.build()` is called, after which it _only_ has getters), or having a the moment be "frozen" as soon as it's instantiated, with all of the setters returning a clone.

For example:

```js
/* BUILDER PATTERN */
var bldr = moment.immutable()
  .hours(5)
  .minutes(30)
  .seconds(25);

bldr.hours();  // throws exception.  builder has no getters

var time = bldr.build();

time.hours(); // 5
time.hours(6); // throws, OR returns a clone

/*  A more explicit variant:  */
var bldr = moment.immutable()
  .setHours(5);

bldr.getHours; // undefined

var time = bldr.build();
time.getHours(); // 5
time.setHours;   // undefined
```

```js
/* VALUE OBJECT */
var time = moment.immutable()   // 00:00:00
  .hours(5)       // new object => 05:00:00
  .minutes(30)    // new object => 05:30:00
  .seconds(25);   // new object => 05:30:25

/*  Same thing, but more efficient:  */
var time2 = moment.immutable(5,30,25); // 05:30:25

time.hours();   // 5
time.hours(6);  // new object => 06:30:25
```

The first example is very java-like, but also leaves little ambiguity about how moments are constructed, and what you can do with them.  It also provides an efficient and intuitive path for constructing new moments with little overhead, and closely follows how most people currently use moment.

I'd be in favor of having setters on built moments throw exceptions, as this behavior would further reduce ambiguity (at the expense of verbosity), and force developers to acknowledge whenever they are creating new moments.

The second example seems like it has a lot more overhead, because we're creating (and discarding) lots of objects.  But, there's a lot of room for optimization.  Modern GC is pretty good; lazy evaluation could help; we could make Moment objects as lightweight as possible (dropping the underlying `Date`), etc.  (For some comparison, this pattern is not unlike the way that strings are handled in JavaScript)

---

I'm a fan of semver, but the browser use-case makes breaking changes problematic, because it'll be impossible to have two versions of moment on the same page without a module-loader.

As previously mentioned, there's a _lot_ of legacy code out there that depends on Moment.  At the very least, we need a release that supports both patterns, to allow for a smooth transition period.

**Comment by alexyuly** [comment]
+1 for composable methods that return new moments. I don't even care if they're immutable. Just give me composition, so I can can chain methods together, or assign the mutated moment in one line (instead of 2).

**Comment by maggiepint** [comment]
@alexyuly why doesn't clone work for you? If you want to compose inline you can do something like:

```js
var a = moment();
var b = a.clone().subtract(1, 'week').startOf('day');
```

Or whatever.

**Comment by alexyuly** [comment]
@maggiepint Well I feel foolish, it looks like you could compose methods all along. Thanks for the tip. Maybe make it a little clearer in the docs?  e.g. http://momentjs.com/docs/#/manipulating/start-of/ <- no mention of any return value.

**Comment by maggiepint** [comment]
Little update - this blog post is MY position on this issue at this time: https://maggiepint.com/2016/06/24/why-moment-js-isnt-immutable-yet/

I don't speak for the rest of the team, but they have seen it, and I'd generally say we're all in a similar place.

**Comment by es6Test** [comment]
@maggiepint great and very valid points

**Comment by ericlau-solid** [comment]
Thanks for this. Immutability is king in redux, so I will switch to js-joda for sure. I may still keep moment for relative time.

Eric Lau - Gmail




On Fri, Jun 24, 2016 at 11:12 AM -0700, "Maggie Pint" <notifications@github.com> wrote:












Little update - this blog post is MY position on this issue at this time: https://maggiepint.com/2016/06/24/why-moment-js-isnt-immutable-yet/



I don't speak for the rest of the team, but they have seen it, and I'd generally say we're all in a similar place.

**Comment by magwo** [comment]
Some typical bugs/confusion caused by mutation: http://stackoverflow.com/questions/26131003/moment-js-start-and-end-of-given-month

**Comment by epeli** [comment]
+1 for this but actually with Redux I have found this to be surprisingly small issue because I don't put Moment instances into the Redux state object at all. Only the date strings or timestamps and I use Moment in the render methods just to convert the date strings to other strings.

I do this mainly because then it is easy to serialize and deserialize the state without any extra work but it also hides the mutation nicely from React and Redux.

**Comment by maggiepint** [comment]
Doing my best to get you an update on this one this weekend. We have a plan, but I need to make sure it's all going to work out before I commit to it. Unfortunately I'm moving from Minneapolis to Seattle next week, so this may slow down for a time, but I want to move it forward.

**Comment by maggiepint** [comment]
Small update in preparation for a bigger update, since I've noticed that that blog post is still getting some  traffic. I am currently re-working moment's build to use Babel. This stages us to bring in @butterflyhug's plugin to the core repository,  and make it officially supported. In the process, we are going to clean up some things about the plugin. @butterflyhug has very graciously been helping us a ton with this. In addition, I'm meeting with John-David Dalton of LoDash (we both work at Microsoft) to discuss strategy when bringing in an official plugin, as he already did it with LoDash functional. Long blog post and RFC coming soon.

**Comment by gabrielmaldi** [comment]
@maggiepint while making frozen-moment officially supported, are you planning on resolving https://github.com/WhoopInc/frozen-moment/issues/20?

I think that this approach to immutability is good enough, but only if every call to `moment()` returns a frozen moment by default, and devs don't have to remember calling `.freeze()` every time.

**Comment by butterflyhug** [comment]
I've finished writing [the draft RFC](https://github.com/butterflyhug/moment-rfcs/blob/frozen-api/0000-immutability.md) and [opened a PR](https://github.com/moment/moment-rfcs/pull/2).

It's quite a long document, in part because there's still a few outstanding questions about how we'll want to deliver this functionality to users and so I'm trying to highlight the pros and cons of various approaches there.  But I think the implementation ideas are pretty solid, so I'd love to hear from folks here if you have reservations about [the proposed user-facing API](https://github.com/butterflyhug/moment-rfcs/blob/frozen-api/0000-immutability.md#new-apis-for-moment-users).

**Comment by maggiepint** [comment]
Executive summary of RFC with our primary questions to be answered: https://maggiepint.com/2016/07/16/immutability-its-happening/

**Comment by jbreckmckye** [comment]
@maggiepint I tried to post a comment on that article, but for some reason it was swallowed. Here's what I wrote:

---

My biggest concern with these changes is that they respond to a disproportionately vocal and expressive part of the community - avoiding the quiet bulwark of ordinary developers who aren't even aware this discussion is taking place.

GitHub's threads are not a microcosm of the wider development community. It suffers the same participation bias as many forums on the web, and is biased towards a certain kind of developer: intellectually engaged with the theory; interested in ideas rather than applications; dare I say socially inclined to rallying around popular trends for the sake of their reputations. These developers are naturally attracted to philosophical cause célèbres like immutability and functional programming, and have closer access to you than any other group. They will be the loudest voice in the room, and they will clamour for this change - but what of the wider world?

The wider world wants Stuff That Works. It wants to know that the library it currently uses will continue receiving updates - bugfixes at least, but ideally small incremental improvements that make their lives better.

If you are intent on this change, you will need to make a very clear case to these people why this change will make their lives better; why they should upgrade; why they needn't worry their legacy projects won't receive indirect bugfixes; what things they will now be able to do that they could not presently. Because the wider world really does not care at all about immutability for its own sake - it cares about applications, not ideas.

I think you should also be very careful to validate the importance of this change in real terms. I would release this change as a plugin or wrapper, and monitor its uptake carefully over several months before I merged it into trunk. If immutability has been over-represented as a niche concern, you will have found a way to satisfy these vocal users without changing the course of the library as a whole.

**Comment by asokani** [comment]
As a first time user I ended here after some time lost with startOf endOf. These methods are surprisingly mutating!

+1 for full immutability

**Comment by AdamHess** [comment]
Maybe another solution is to make it **painfully obvious that objects in momentjs are mutable**. The documentation does mention it in the cloning section, but it is NOT salient enough.  

Another solution,If you want mutability use Object Oriented concepts and create object creation using the **NEW** keyword vs the moment() factory pattern.

**Comment by igl** [comment]
> The wider world wants Stuff That Works.
> ...you will need to make a very clear case to these people why this change will make their lives better

Everybody i have witnessed being new to moment falls into the trap of "moments" being mutated.
Even after 3 years of using moment i still have to tell myself "oh shit you are using .startOf here .. better check twice if you need a copy".

The current behaviour is not intuitive and this is a long time overdue.

**Comment by mull** [comment]
I like @AdamHess' idea about choosing whether you're looking for OOP or immutability.

My two cents about why we get confused is this: return values. If `moment.add('1', 'days')` was returning undefined, like mutable functions in JS generally do, then the confusion would go away. Returning the same object, for me at least, implies I have a new copy. Of course, that would break chainability.

**Comment by andyfleming** [comment]
I think there is a low likelihood of hitting memory use issues for more developers (except for special use cases). Moment's mutability has already bitten me though. Dates should be treated as values like a string or number.

+1 for immutable by default

**Comment by andyfleming** [comment]
This same issue exists for PHP, by the way. Do you want to be like PHP?! 😆 

In PHP, they solve it by providing `DateTimeImmutable` (in addition to the normal `DateTime`).

If we don't change the default behavior to be immutable, we should at least consider a first-class alternate API like `imoment`/`momenti` (or whatever). I would literally always use that though (over the mutable API) and I would hope that any other library I'd use would be using the immutable version/API too.

**Comment by ngerritsen** [comment]
I vote for immutability too, I'm willing to help with implementing if that's the plan. I already ran into mutability issues when doing addition and substraction, what makes it more confusing even, is that these methods do return the instance because of chaining.

**Comment by ngerritsen** [comment]
By the way how about cloning durations? What's the best way of doing that, couldn't find any that did not feel hacky.

**Comment by butterflyhug** [comment]
@ngerritsen I believe the best available option is `moment.duration(existingDuration)`.

Re: implementation, #3548 is still an active PR.  Hopefully there isn't much code-level work left, but it never hurts to have more eyes to validate big changes.  We also need to work on documentation and etc before we can do a major version bump, which will be necessary in order to release a change like this.  If you'd like to help with any of this list, I'm sure we'd appreciate it.  😀

**Comment by fabiosussetto** [comment]
Just spent an hour trying to identify a subtle bug caused by .startOf() mutating the original date... Thanks for the hard work, momentjs did a great job showing how to build an excellent date lib for JS, but I'm switching to date-fns  because of the very subtle nature of this kind of bugs.
For what it's worth, lodash has already followed a more FP approach with lodash/fp.

**Comment by noway** [comment]
While moment.frozen namespace is in the works, I'd recommend --as previous poster suggested-- to just use [date-fns](https://date-fns.org/).

**Comment by igl** [comment]
Just fixed another bug due to mutable moments 🎉

**Comment by sandstrom** [comment]
Tangentially related, it would be awesome if 3.0 could move to ES6 classes:

```js
let mom1 = new Moment();
let mom2 = Moment.parse('2019-03-01T14:55');
// etc
```

Such a move could also guide the immutability discussion. I'd say all methods should be immutable with one exception. A method called `.set('minute/hour/year/etc', 18)`.

**Comment by QuantumInformation** [comment]
Time to mute this thread

**Comment by igl** [comment]
@alancnet moment will probably never be changed. It's too much of a change and there is enough push back from users who learned to embrace the current behaviour.

Check out their new project: [luxon](https://github.com/moment/luxon)
It's very nice, modern and should perform way better than moment-immutable which just wraps everything in `.clone()` calls.

**Comment by stevemao** [comment]
For those who'd like to transition from momentjs to a modern approach, please check out https://github.com/you-dont-need/You-Dont-Need-Momentjs

**Comment by sandstrom** [comment]
@alamothe That question is clearly answered already: https://moment.github.io/luxon/docs/manual/moment.html

**Comment by mj1856** [comment]
Closing this.  As others have pointed out, use Luxon if you want a mostly immutable API.  Thanks.
